src/server/handlers/launchpad.js

```diff
@@ -122,107 +143,137 @@ const responseError = (res, error) => {
   }
 };
 
-const makeSnapName = url => {
-  return createHash('md5').update(url).digest('hex');
+const checkStatus = (response) => {
+  if (response.statusCode !== 200) {
+    let body = response.body;
+    if (typeof body !== 'object') {
+      try {
+        body = JSON.parse(body);
+      } catch (e) {
+        logger.info('Invalid JSON received', e, body);
+        throw new PreparedError(500, RESPONSE_GITHUB_OTHER);
+      }
+    }
+    switch (body.message) {
+      case 'Not Found':
+        // snapcraft.yaml not found
+        throw new PreparedError(404, RESPONSE_GITHUB_NOT_FOUND);
+      case 'Bad credentials':
+        // Authentication failed
+        throw new PreparedError(401, RESPONSE_GITHUB_AUTHENTICATION_FAILED);
+      default:
+        // Something else
+        logger.info('GitHub API error:', response.statusCode, body);
+        throw new PreparedError(500, RESPONSE_GITHUB_OTHER);
+    }
+  }
+  return response;
 };
 
-const getSnapcraftYaml = (req, res, callback) => {
+const checkAdminPermissions = (req) => {
+  if (!req.session || !req.session.token) {
+    return Promise.reject(new PreparedError(401, RESPONSE_NOT_LOGGED_IN));
+  }
+  const token = req.session.token;
+
   const repositoryUrl = req.body.repository_url;
   const parsed = parseGitHubUrl(repositoryUrl);
   if (parsed === null || parsed.owner === null || parsed.name === null) {
     logger.info(`Cannot parse "${repositoryUrl}"`);
-    return res.status(400).send(RESPONSE_GITHUB_BAD_URL);
+    return Promise.reject(new PreparedError(400, RESPONSE_GITHUB_BAD_URL));
   }
 
-  const uri = `/repos/${parsed.owner}/${parsed.name}/contents/snapcraft.yaml`;
+  const uri = `/repos/${parsed.owner}/${parsed.name}`;
+  const options = {
+    headers: { 'Authorization': `token ${token}` },
+    json: true
+  };
+  logger.info(`Checking permissions for ${parsed.owner}/${parsed.name}`);
+  return requestGitHub.get(uri, options)
+    .then(checkStatus)
+    .then(response => {
+      if (!response.body.permissions || !response.body.permissions.admin) {
+        throw new PreparedError(401, RESPONSE_GITHUB_NO_ADMIN_PERMISSIONS);
+      }
+      return [parsed.owner, parsed.name, token];
+    });
+};
+
+const makeSnapName = (url) => {
+  return createHash('md5').update(url).digest('hex');
+};
+
+const getSnapcraftYaml = (owner, name, token) => {
+  const uri = `/repos/${owner}/${name}/contents/snapcraft.yaml`;
   const options = {
     headers: {
-      'Authorization': `token ${req.session.token}`,
+      'Authorization': `token ${token}`,
```

Comment:
We add the 'Authorization' header to every GH request, so maybe `requestGitHub` could do it? It probably can't take token from session on its own, but we can just pass it as param or part of options (not building whole header value each time).